<h1>Part 3a - Chessboard Detection: Train + Predict with Model Results</h1>

<h2>Board_Detection Object</h2>

 A labeler, trainer and a detector for screenshots. Labeler is implemented in gcb_utils while trainer and detector are interfaces into PyTorch implementation of Yolov5 (https://github.com/ultralytics/yolov5, see https://pjreddie.com/media/files/papers/yolo.pdf for Yolo.) Main methods are update_labels, train and predict.<br>

In [1]:
from board_detection import Board_Detection
import gcb_utils.gcb_utils as gcb_utils

<h2>Initialize the Model</h2>

Note that we're not specifying model weights, yet. It's an optional parameter at this point.

In [2]:
model_dir = 'data/model/board-final_weight/'

board_detector = Board_Detection(model_dir)

No model weight file specified. You can train a new model using Board_Detection.train.


<h2>Update Labels - Board_Detection.update_labels()</h2>

Although demonstrated in Part 2a, I am including the update_labels() function of Board_Detection here for completeness.      

In [3]:
#please uncomment the below for documentation
#print(help(board_detector.update_labels))

In [ ]:
screenshot_data_path = 'data/raw/screenshots'
screenshot_labels_fname = 'data/model/screenshot_boundboxes.csv'
SCREENSHOT_LABEL_COLUMNS = ['fname', 'height_pxl','width_pxl','label','x_min_pxl','y_min_pxl'    ,'x_max_pxl','y_max_pxl', 'HumCheck-YN']
update_fn = gcb_utils.screenshot_height_width_update

board_detector.update_labels(screenshot_data_path, screenshot_labels_fname, SCREENSHOT_LABEL_COLUMNS, update_fn=update_fn, update_fn_kwargs={'screenshot_path':screenshot_data_path})

<h2>Training-i: 1-Epoch Run in the Notebook</h2>

<li>The function .train() uses yolov5.detect() for training. (I've tried as much as I can to not modify the yolov5 directory, however please note the   __init__.py file I've inserted into the yolov5 directory) <br><br>

<li>Following is a-1 epoch training of the dataset for demonstration purposes. Actual training took around 39 hours on a Macbook Air M1 where each epoch took approximately 4 minutes.

In [5]:
#please uncomment the below for documentation
#print(help(board_detector.train))

In [6]:
yaml_path_or_dict = 'data/model/scr_data_cli.yaml' #file pointing to inputs and class labels.
init_weights_dir = 'yolov5'
init_weights_fname = 'yolov5s.pt' #weights provided yolov5 pretrained on COCO
train_project_dir = 'data/model/board-train'
train_project_name = 'exp'
epochs = 1


board_detector.train(yaml_path_or_dict, 
                     init_weights_dir, 
                     init_weights_fname, 
                     train_project_dir, 
                     train_project_name, 
                     epochs=epochs)

train: weights=yolov5/yolov5s.pt, cfg=, data=data/model/scr_data_cli.yaml, hyp=yolov5/data/hyps/hyp.scratch.yaml, epochs=1, batch_size=16, imgsz=1440, rect=False, resume=False, nosave=False, noval=False, noautoanchor=False, evolve=None, bucket=, cache=True, image_weights=False, device=cpu, multi_scale=False, single_cls=False, adam=False, sync_bn=False, workers=8, project=data/model/board-train, name=exp, exist_ok=False, quad=False, linear_lr=False, label_smoothing=0.0, patience=100, freeze=0, save_period=-1, local_rank=-1, entity=None, upload_dataset=False, bbox_interval=-1, artifact_alias=latest


Namespace(adam=False, artifact_alias='latest', batch_size=16, bbox_interval=-1, bucket='', cache=True, cfg='', data='data/model/scr_data_cli.yaml', device='cpu', entity=None, epochs=1, evolve=None, exist_ok=False, freeze=0, hyp=PosixPath('yolov5/data/hyps/hyp.scratch.yaml'), image_weights=False, imgsz=1440, label_smoothing=0.0, linear_lr=False, local_rank=-1, multi_scale=False, name='exp', noautoanchor=False, nosave=False, noval=False, patience=100, project='data/model/board-train', quad=False, rect=False, resume=False, save_period=-1, single_cls=False, sync_bn=False, upload_dataset=False, weights='yolov5/yolov5s.pt', workers=8)
github: 

YOLOv5 🚀 v6.0-114-ga4207a2 torch 1.9.0 CPU

hyperparameters: lr0=0.01, lrf=0.1, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=0.05, cls=0.5, cls_pw=1.0, obj=1.0, obj_pw=1.0, iou_t=0.2, anchor_t=4.0, fl_gamma=0.0, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0
TensorBoard: Start with 'tensorboard --logdir data/model/board-train', view at http://localhost:6006/


⚠️ YOLOv5 is out of date by 27 commits. Use `git pull` or `git clone https://github.com/ultralytics/yolov5` to update.
Weights & Biases: run 'pip install wandb' to automatically track and visualize YOLOv5 🚀 runs (RECOMMENDED)


Overriding model.yaml nc=80 with nc=1

                 from  n    params  module                                  arguments                     
  0                -1  1      3520  models.common.Conv                      [3, 32, 6, 2, 2]              
  1                -1  1     18560  models.common.Conv                      [32, 64, 3, 2]                
  2                -1  1     18816  models.common.C3                        [64, 64, 1]                   
  3                -1  1     73984  models.common.Conv                      [64, 128, 3, 2]               
  4                -1  2    115712  models.common.C3                        [128, 128, 2]                 
  5                -1  1    295424  models.common.Conv                      [128, 256, 3, 2]              
  6                -1  3    625152  models.common.C3                        [256, 256, 3]                 
  7                -1  1   1180672  models.common.Conv                      [256, 512, 3, 2]             


1 epochs completed in 0.077 hours.

Validating data/model/board-train/exp/weights/best.pt...
Fusing layers... 
Model Summary: 213 layers, 7012822 parameters, 0 gradients


Optimizer stripped from data/model/board-train/exp/weights/last.pt, 14.9MB
Optimizer stripped from data/model/board-train/exp/weights/best.pt, 14.9MB


               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|██████


                 all         18         24    0.00205       0.25    0.00114   0.000163
Results saved to data/model/board-train/exp


<h2>Training-ii: Actual Training and Results:</h2>
    
<h3> Training parameters:</h3>

The dataset contained 64 images for training and 18 for validation. <br>
    
I have run model on the command line at yolov5 root with: (the environment can be found at requirements.txt at the repo root)  <br>
    python train.py --img 1440  --epochs 1500 --data ../data/model/scr_data.yaml --weights yolov5s.pt --cache --device cpu

Going over the command line options used:
    <li>--img 1440 -> size in pixels for the longer side of image
    <li>--epochs 1500 -> Time cost of early termination was high - I picked a large number.
    <li>--data -> indicates the yaml file created for the dataset (see Part 2a)
    <li>--weights -> initial pretrained weights from which to start
    <li>for any other options (e.g. patience, please refer to yolov5.train.py, or to a narrower extent, Board_Detection.train)


        
<h3> Results:</h3>
Final results on an Apple MacBook Air (2020) M1 with 16GB of memory are as follows. Each epoch took approximately 4 minutes. Below, I show the best (498) and final (598) epochs. The final epoch is less than the indicated 1500 due to default patience parameter of 100. If an epoch survives as best after 100 steps, training is stopped.  
    
![](z_markdown_jpgs/Board-Best+FinalEpochs.png) <br>

<li>Over the 18 images used for validation, mean Average Precision (mAP) at an Intersection Over Union (IOU) ratio of 0.5 was 0.987. Over a range of IOU's from 0.5 to 0.95 mAP stood at 0.953. Also, please note that precision is 0.96 and recall is >0.995 in the validation set. The precision-recall graph from yolov5 is presented below.<br> <br>
    
![](z_markdown_jpgs/BoardPR_curve.png) <br>
    
<li>Although, there is a chance that model could be overfit due to these high figures, unfortunately I did not have the chance to experiment with the model given the high time cost of training. The results from test images however, show that the fit is still functional out of the sample for the purpose of detecting chessboards. <br><br>
    
<li>Below I also present a validation image mosaic from yolov5 output.
 
 ![](z_markdown_jpgs/Boardval_batch0_labels.jpg) <br>
    Validation Set - Labels <br><br>
    
 ![](z_markdown_jpgs/Boardval_batch0_pred.jpg) <br>
    Validation Set - Predictions <br><br>
    


<h2>Detection (.predict): Demonstration using Final Model and Test Images</h2>
 
.predict() outputs board predictions of files in source_dir with bounding boxes using the yolov5.detect function. The function also returns a pandas DataFrame that contains any labels it found for the image set.

Below, I will use the final model weights from the longer run and apply them on 10 test images which did not participate in training. Note that model_dir has to include hubconf.py. Function help is available below.

In [7]:
#please uncomment the below for documentation
#print(help(board_detector.predict))

In [8]:
model_dir = 'data/model/board-final_weight'
model_weight_fname = 'board_best.pt' 
source_dir = 'data/raw/screenshots/test'
imgsz = (1440, 1440) #Square shape image size 
project_dir = 'data/model/board-detect'
project_name = 'det'
conf_thres = 0.10

board_detector = Board_Detection(model_dir, model_weight_fname)
prediction = board_detector.predict(model_dir, 
                                    model_weight_fname, 
                                    source_dir, 
                                    imgsz, 
                                    project_dir, 
                                    project_name,  
                                    conf_thres, 
                                    save_txt = True, 
                                    save_conf = True)

YOLOv5 🚀 v6.0-114-ga4207a2 torch 1.9.0 CPU

Fusing layers... 
Model Summary: 213 layers, 7012822 parameters, 0 gradients


image 1/10 /Users/artun/Desktop/Springboard/springboard/Get-Chess-Board/data/raw/screenshots/test/Screen Shot 2021-12-01 at 7.44.47 PM.jpg: 928x1440 4 Chessboards, Done. (0.435s)
image 2/10 /Users/artun/Desktop/Springboard/springboard/Get-Chess-Board/data/raw/screenshots/test/Screen Shot 2021-12-01 at 7.45.27 PM.jpg: 928x1440 1 Chessboard, Done. (0.407s)
image 3/10 /Users/artun/Desktop/Springboard/springboard/Get-Chess-Board/data/raw/screenshots/test/Screen Shot 2021-12-01 at 7.46.17 PM.jpg: 928x1440 3 Chessboards, Done. (0.408s)
image 4/10 /Users/artun/Desktop/Springboard/springboard/Get-Chess-Board/data/raw/screenshots/test/Screen Shot 2021-12-01 at 7.55.43 PM.jpg: 928x1440 2 Chessboards, Done. (0.409s)
image 5/10 /Users/artun/Desktop/Springboard/springboard/Get-Chess-Board/data/raw/screenshots/test/Screen Shot 2021-12-01 at 7.56.39 PM.jpg: 928x1440 1 Chessboard, Done. (0.407s)
image 6/10 /Users/artun/Desktop/Springboard/springboard/Get-Chess-Board/data/raw/screenshots/test/Screen Sh

yolov5 detection results are saved in: /Users/artun/Desktop/Springboard/springboard/Get-Chess-Board/data/model/board-detect/det


In [9]:
#output is a pd.DataFrame
prediction

fname label  x_min_norm  y_min_norm  \
0  Screen Shot 2021-12-01 at 7.44.47 PM     0    0.295139    0.480556   
1  Screen Shot 2021-12-01 at 7.44.47 PM     0    0.724306    0.673889   
2  Screen Shot 2021-12-01 at 7.44.47 PM     0    0.648264    0.830000   
0  Screen Shot 2021-12-01 at 7.46.17 PM     0    0.747569    0.855556   
1  Screen Shot 2021-12-01 at 7.46.17 PM     0    0.502431    0.506111   
0  Screen Shot 2021-12-01 at 8.03.09 PM     0    0.395833    0.855000   
1  Screen Shot 2021-12-01 at 8.03.09 PM     0    0.260417    0.496111   
0  Screen Shot 2021-12-01 at 7.55.43 PM     0    0.481597    0.850556   

   x_max_norm  y_max_norm  confidence  
0    0.547222    0.858889    0.863039  
1    0.123611    0.078889    0.877888  
2    0.071528    0.115556    0.934162  
0    0.071528    0.111111    0.932372  
1    0.536806    0.861111    0.943299  
0    0.052778    0.090000    0.920412  
1    0.408333    0.630000    0.934135  
0    0.057639    0.090000    0.918741

<h3>Board Detection Results</h3>

Let's take a look at the board detection output by yolov5 on images not previously seen:

  Some test outputs with correct classification:
![](z_markdown_jpgs/Board_test_pred_ScreenShot2021-12-01at7.45.27PM.jpg) <br>
![](z_markdown_jpgs/Board_test_pred_ScreenShot2021-12-01at7.56.39PM.jpg) <br><br><br>

  However, although the model seems to perform well in test, not all classifications were correct. <br>
  Here is an image with a False Positive (please note the partial chessboard and the smaller square with confidence values of 0.80):
![](z_markdown_jpgs/Board_test_pred_ScreenShot2021-12-01at7.44.47PM.jpg) <br>

  Here is also an image with a False Negative (please note the chessboard not labeled):
![](z_markdown_jpgs/Board_test_pred_ScreenShot2021-12-01at7.55.43PM.jpg) <br>